doc/adr/0006-protocol-naming-conventions.md

```diff
@@ -0,0 +1,56 @@
+# 6. protocol-naming-conventions
+
+Date: 2021-05-08
+
+## Status
+
+Under Review
+
+## Context
+
+This document describes naming conventions for these protocol components:
+
+* message subject
+* stream name
+* reply to
+* queue name
+* durable name
+* account name
+
+## Prior Work
+
+Currently the NATS Docs regarding [protocol convention](https://docs.nats.io/nats-protocol/nats-protocol#protocol-conventions) says this:
+
+> Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace. All ascii alphanumeric characters except spaces/tabs and separators which are "." and ">" are allowed. Subject names can be optionally token-delimited using the dot character (.), e.g.:
+A subject is comprised of 1 or more tokens. Tokens are separated by "." and can be any non space ascii alphanumeric character. The full wildcard token ">" is only valid as the last token and matches all tokens past that point. A token wildcard, "*" matches any token in the position it was listed. Wildcard tokens should only be used in a wildcard capacity and not part of a literal token.
```

Comment:
> should only be used in a wildcard capacity and not part of a literal token

They should but we support them being just characters in a token, that is, as literal as opposed to a wildcard. We had users in the past reporting that we were not treating them properly and had to fix that, and the server (at least) is supposed to recognize "foo*.bar" as a literal subject with 2 tokens, one `foo*` and one `bar`.
But it is understood that in this case `foo*` is just a string and has no wildcard semantic, which some may think that it would "match"  any token that starts with `foo`.

Reply:
Is `*` a regular character if it is followed by `.`, `*` or `>` or what would the rules? The prior work is currently the published doc.

Reply:
Both `*` and `>` are regular characters if they are not single character forming a token. So in theory `foo***bar.baz>.bat` is a valid subject.

Reply:
@derekcollison  / @ColinSullivan1 care to comment? I've been repeatedly told that the docs are the standard. I imagine this is not an issue with newer clients like java because the users are not aware of this. I have no idea what the go client currently supports. I'm guessing C++ supports this more lenient specification.  I wonder what other clients support.
Since this is more lenient than the documentation, it is certainly more adoptable than the other way.
We could note specifically that existing clients that support this behavior will continue to allow this, but interoperability between clients strongly suggest adherence to the strict definition, and that in the future the non-strict standard may be deprecated. We could also just leave the docs as is - do we know of users that are in fact using it and how worried are we that we break them?

Reply:
I care quite a bit if we are potentially breaking existing deployments.

Reply:
I am fine with above since you mention "should only be used", not "must only be used".